src/fparser/two/tests/fortran2003/test_loop_control_r830.py

```diff
@@ -0,0 +1,116 @@
+# notice, this list of conditions and the following disclaimer.
+
+# 2. Redistributions in binary form must reproduce the above copyright
+# notice, this list of conditions and the following disclaimer in the
+# documentation and/or other materials provided with the distribution.
+
+# 3. Neither the name of the copyright holder nor the names of its
+# contributors may be used to endorse or promote products derived from
+# this software without specific prior written permission.
+
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+# A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+# HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+# LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+"""Test Fortran 2003 rule R830 : This file tests the support for the
+loop-control rule.
+
+"""
+import pytest
+from fparser.two.Fortran2003 import Loop_Control
+from fparser.two.utils import NoMatchError
+
+
+def test_start_end_space_while():
+    """Test that there is a match if the string contains white space at
+    the start and end and that the tostr() output is as expected. Also
+    test matching to the while form of this rule.
+
+    """
+    result = Loop_Control(" while (.true.) ")
+    assert isinstance(result, Loop_Control)
+    assert str(result) == "WHILE (.TRUE.)"
+
+
+def test_delim():
+    """Test that there is a match if the string contains an optional
+    delimiter at the start and that the tostr() output is as expected.
+
+    """
+    result = Loop_Control(" , while (.true.) ")
+    assert isinstance(result, Loop_Control)
+    assert str(result) == ", WHILE (.TRUE.)"
+
+
+def test_repmap():
+    """Test matching when the while logical expresssion contains
+    brackets. This tests the use of the string_replace_map() function.
+
+    """
+    result = Loop_Control(" , while (((a .or. b) .and. (c .or. d))) ")
+    assert isinstance(result, Loop_Control)
+    assert str(result) == ", WHILE (((a .OR. b) .AND. (c .OR. d)))"
+
+
+def test_counter():
+    """Test matching to the counter form of this rule and that the tostr()
+    output is as expected.
+
+    """
+    # Lower-bound and upper-bound only
+    result = Loop_Control("idx = start,stop")
+    assert isinstance(result, Loop_Control)
+    assert str(result) == "idx = start, stop"
+    # Lower-bound, upper-bound and step
+    result = Loop_Control("idx = start,stop,step")
```

Comment:
Please could you extend this by testing when one or more of the start, stop or step values are actually expressions involving parentheses.

Reply:
Extended.